### PR TITLE
Align WritableStream structure with ReadableStream structure

### PR DESCRIPTION
**Pull request by tyoshino**
- Add WritableStreamDefaultWriter
- Add WritableStreamDefaultController
- Add lock related methods
- Update reader.pipeTo() to use the new WritableStream classes

### CODE REVIEW

**Comment by tyoshino** [comment]
This is WIP. I'll create some issues to discuss design decisions.

**Comment by tyoshino** [comment]
- Synchronous (or asynchronous) state probing on WritableStream https://github.com/whatwg/streams/issues/463

**Comment by tyoshino** [comment]
- [WritableStream] Shouldn't the state after writer.abort() be closed? https://github.com/whatwg/streams/issues/464

**Comment by tyoshino** [comment]
- Synchronous queue draining for WritableStream https://github.com/whatwg/streams/issues/465

**Comment by tyoshino** [comment]
I also started updating TransformStream. It should reveal issues of WritableStream if any.

File: reference-implementation/lib/writable-stream.js [hunk 1]
```diff
@@ -7,222 +7,462 @@ const { DequeueValue, EnqueueValueWithSize, GetTotalQueueSize, PeekQueueValue }
...
+    return IsWritableStreamLocked(this);
+  }
+
+  getWriter() {
+    if (IsWritableStream(this) === false) {
+      throw new TypeError('WritableStream.prototype.getWriter can only be used on a WritableStream');
+    }
+
+    return AcquireWritableStreamDefaultWriter(this);
+  }
```
Comment (domenic) [inline reference-implementation/lib/writable-stream.js]: Although I know there is a thread on abort(), for now (not having looked at that thread in detail yet) I think we should keep abort() as a lock-abort-unlock, similar to how cancel() on readable streams is lock-cancel-unlock.

Comment (domenic) [inline reference-implementation/lib/writable-stream.js]: Oh, I see, you did that in the follow-up commit; nevermind!

File: reference-implementation/lib/writable-stream.js [hunk 2]
```diff
@@ -35,6 +35,18 @@ class WritableStream {
     return IsWritableStreamLocked(this);
   }
 
+  abort(reason) {
+    if (IsWritableStream(this) === false) {
+      throw Promise.reject(new TypeError('WritableStream.prototype.abort can only be used on a WritableStream'));
+    }
+
+    if (IsWritableStreamLocked(this) === true) {
+      throw Promise.reject(new TypeError('Cannot abort a stream that already has a reader'));
```
Comment (domenic) [inline reference-implementation/lib/writable-stream.js]: return, not throw. Need tests to catch this.

**Comment by domenic** [comment]
Wow, thank you so much for taking this on. This is so great.

I wonder what strategy you have in mind for doing this work. For example, IMO it would be valuable to fix writable streams in the ways we *know* we want to change, and get the spec + tests for them done. All of the stuff in the first two commits seems great to me. Then we could work on transform streams and other less-obvious fixes afterward. What do you think?

Another thing we need to do is convert the existing tests to web platform tests :(. Don't necessarily worry about that for now, but maybe if writing new tests try to keep them in WPT format and directories.

I'll go off to comment on the other issues now!

File: reference-implementation/lib/writable-stream.js (hunk 2 above)
Comment (tyoshino) [inline reference-implementation/lib/writable-stream.js]: Thanks. Fixed and added a test.

File: reference-implementation/lib/writable-stream.js (hunk 1 above)
Comment (tyoshino) [inline reference-implementation/lib/writable-stream.js]: Yeah!

**Comment by tyoshino** [comment]
> it would be valuable to fix writable streams in the ways we know we want to change,

Yeah. There some experimental features are implemented in this PR, but they'll be reverted and this PR would just include only what we agreed e.g. writer introduction.

**Comment by tyoshino** [comment]
All tests pass except for skipped ones which are related to the locking semantics. I'll fix them and then tweak it / fix style issues.

File: reference-implementation/lib/writable-stream.js [hunk 3]
```diff
@@ -578,14 +578,15 @@ class WritableStreamDefaultController {
     const startResult = InvokeOrNoop(underlyingSink, 'start', [this]);
     Promise.resolve(startResult).then(
       () => {
+        throw new assert.AssertionError('ss');
```
Comment (LinusU) [inline reference-implementation/lib/writable-stream.js]: Did you mean to commit this line?

Comment (tyoshino) [inline reference-implementation/lib/writable-stream.js]: Thanks. Removed

**Comment by tyoshino** [comment]
Moved to https://github.com/whatwg/streams/pull/488 to keep all the logs while making the PR rebased and squashed for easier review.

**Pull request closed without merging**